build: Add meson build

hadess opened the pull request:
This simplifies the installation of necessary data files for the thumbnailer to work, working around this long-standing RFE: https://github.com/rust-lang/cargo/issues/2729

Ideally, meson would also be used to compile the binary instead of Cargo, but meson doesn't easily support external crates: https://github.com/mesonbuild/meson/issues/2173

hadess committed:
build: Add meson build

This simplifies the installation of necessary data files for the
thumbnailer to work, working around this long-standing RFE:
https://github.com/rust-lang/cargo/issues/2729

Ideally, meson would also be used to compile the binary instead of
Cargo, but meson doesn't easily support external crates:
https://github.com/mesonbuild/meson/issues/2173

hadess commented:
When this lands, I can retire `gnome-nds-thumbnailer`, thanks :)

MateusRodCosta commented:
Hi, I will look into this over this weekend.

> When this lands, I can retire `gnome-nds-thumbnailer`, thanks :)

Apparently gnome-nds-thumbnailer got removed from Fedora a while ago...
But, I would guess being in official repos instead of simply COPR would be important to be considered a real replacement.

MateusRodCosta commented on meson.build:
```diff
@@ -0,0 +1,51 @@
+    install_dir: get_option('datadir') / 'mime/packages',
+)
+
+gnome.post_install(
+    update_mime_database: true
+)
```
Can this be done conditionally so I can re-use the meson build on the RPM spec?

hadess replied:
Fixed in the additional commit, let me know.

MateusRodCosta reviewed (changes requested):
I'm a bit unsure on how to integrate this with [the RPM spec](https://pagure.io/bign-handheld-thumbnailer-spec/blob/main/f/bign-handheld-thumbnailer.spec).

Sure, I might be able to run some macro that triggers meson (not sure which yet) and that will save me from having to manually call `install` to install the .thumbnailer file and the mime types. But will that still keep the licenses generation macros working?

Also it seems that the meson file calls the update mime database command unconditionally, which is fine if it's only used for manual installing, but it could be a problem if I want to integrate with the spec, as the mime type database update will be triggered on user's system automatically.

hadess committed:
build: Make update_mime_database optional

As it needs to be disabled for the installation process when building
distribution packages.

hadess commented:
> Apparently gnome-nds-thumbnailer got removed from Fedora a while ago...

About 18 months ago, not certain why.

> Sure, I might be able to run some macro that triggers meson (not sure which yet) and that will save me from having to manually call `install` to install the .thumbnailer file and the mime types. But will that still keep the licenses generation macros working?

meson just runs cargo. We could also make meson generate `LICENSE` similarly to `%{cargo_license_summary}` or `LICENSE.dependencies` like `%{cargo_license}` if you want, or you could keep running them manually in the spec file.

You can use:
- `%meson` instead of `%cargo_build`
- `%meson_install` instead of `%cargo_install`.

MateusRodCosta reviewed (approved):
lgtm, though I would initially expect `update_mime_database` to be opt-in instead of opt-out.
But, since we only don't want to do it for packaging, I think it makes sense to be opt-out.

MateusRodCosta commented:
> About 18 months ago, not certain why.

From a search on the Fedora Devel list apparently they were removing inactive packages and orphaning their packages, and it seems gnome-nds-thumbnailer got caught into this, see [here](https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/message/JUCF3UJ7XAL3XB7HC5RT57PSCPMQTQMD/).

> meson just runs cargo. We could also make meson generate LICENSE similarly to %{cargo_license_summary} or LICENSE.dependencies like %{cargo_license} if you want, or you could keep running them manually in the spec file.
> 
> You can use:
> 
>     %meson instead of %cargo_build
>     %meson_install instead of %cargo_install.

Makes sense, I will tag a new release soonish and then update the spec and update the COPR.

> When this lands, I can retire `gnome-nds-thumbnailer`, thanks :)

I believe that this means that you plan to archive the gnome-nds-thumbnailer repo, right?

It could be a good idea to link to this repo and mention that it not only covers Nintendo DS .nds files but also several Nintendo 3DS file formats as well. Mentioning that it's written in Rust is optional.
I guess this would also point distros and users that still care about having .nds thumbnailing to an appropriate replacement.
